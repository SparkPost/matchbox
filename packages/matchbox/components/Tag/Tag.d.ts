export default Tag;
declare const Tag: any;
