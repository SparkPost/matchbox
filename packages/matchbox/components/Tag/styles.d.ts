export function tagBase(props: any): string;
export function closeBase(): string;
export function content(): string;
export function tagColor(props: any): string;
export function closeColor(props: any): string;
