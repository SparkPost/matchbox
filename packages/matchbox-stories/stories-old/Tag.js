import React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';
import { action } from '@storybook/addon-actions';
import { StoryContainer } from './helpers';

import { Panel, Tag } from '../src';

storiesOf('Tag', module)
  .addDecorator((getStory) => (
    <StoryContainer>{ getStory() }</StoryContainer>
  ))
  .add('Default',
  withInfo()(() => (
    <Tag>domain.com</Tag>
  )))
  .add('With remove action',
  withInfo()(() => (
    <Tag onRemove={action('Tag Remove')}>domain.com</Tag>
  )))
  .add('Colors',
  withInfo()(() => (
    <div>
      <Tag onRemove={action('Tag Remove')} color='orange'>domain.com</Tag>{' '}
      <Tag onRemove={action('Tag Remove')} color='blue'>domain.com</Tag>{' '}
      <Tag onRemove={action('Tag Remove')} color='yellow'>domain.com</Tag>{' '}
      <Tag onRemove={action('Tag Remove')} color='red'>domain.com</Tag>{' '}
      <Tag onRemove={action('Tag Remove')} color='navy'>domain.com</Tag>{' '}
      <Tag onRemove={action('Tag Remove')} color='purple'>domain.com</Tag>{' '}
    </div>
  )));
