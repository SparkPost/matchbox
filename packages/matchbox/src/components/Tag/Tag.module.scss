@import '../../styles/config';

.Tag {
  display: inline-block;
  white-space: nowrap;
}

.hasRemove {
  .Content {
    border-radius: border-radius(large) 0 0 border-radius(large);
  }
}

.orange {
  .Content {
    background: color(orange, dark);
    color: color(gray, 10);
  }

  .Close, a.Close {
    background: color(orange, dark);
    color: color(gray, 9);

    &:hover, &:active {
      color: color(gray, 10);
      background: darken(color(orange, dark), 5);
    }
  }
}

.blue {
  .Content {
    background: color(blue, dark);
    color: color(gray, 10);
  }

  .Close, a.Close {
    background: color(blue, dark);
    color: color(gray, 9);

    &:hover, &:active {
      color: color(gray, 10);
      background: darken(color(blue, dark), 5);
    }
  }
}

.yellow {
  .Content {
    background: color(yellow, light);
    color: color(gray, 1);
  }

  .Close, a.Close {
    background: color(yellow, light);
    color: color(gray, 1);

    &:hover, &:active {
      color: color(gray, 0);
      background: darken(color(yellow, light), 5);
    }
  }
}

.red {
  .Content {
    background: color(red);
    color: color(gray, 10);
  }

  .Close, a.Close {
    background: color(red);
    color: color(gray, 9);

    &:hover, &:active {
      color: color(gray, 10);
      background: color(red, dark);
    }
  }
}

.navy {
  .Content {
    background: color(navy);
    color: color(gray, 10);
  }

  .Close, a.Close {
    background: color(navy);
    color: color(gray, 9);

    &:hover, &:active {
      color: color(gray, 10);
      background: color(navy, dark);
    }
  }
}

.purple {
  .Content {
    background: color(purple);
    color: color(gray, 10);
  }

  .Close, a.Close {
    background: color(purple);
    color: color(gray, 9);

    &:hover, &:active {
      color: color(gray, 10);
      background: color(purple, dark);
    }
  }
}

.Content {
  display: inline-block;
  vertical-align: middle;
  padding: rem(1) rem(7) rem(1);
  border-radius: border-radius(large);
  background: color(gray, 8);

  font-size: font-size(300);
  line-height: rem(22);
}

.Close, a.Close {
  display: inline-block;
  vertical-align: middle;
  padding: 0 rem(6);
  border-radius: 0 border-radius(large) border-radius(large) 0;
  background: color(gray, 8);

  color: color(gray, 4);
  line-height: rem(24);
  transition: 0.15s;

  &:hover, &:active {
    color: color(gray, 1);
    background: color(gray, 7);
  }

  & > * {
    margin-top: -0.2em;
  }
}
