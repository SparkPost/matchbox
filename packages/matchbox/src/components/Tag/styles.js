import { tokens } from '@sparkpost/design-tokens';

export const tagBase = props => `
  display: inline-flex;
  align-items: center;
  white-space: nowrap;
  padding: 1px ${tokens.spacing_300};
  padding-right: ${props.hasRemove ? '1px' : tokens.spacing_300};
  border-radius: ${tokens.borderRadius_pill};
  line-height: ${tokens.lineHeight_200};
  font-size: ${tokens.fontSize_200};
  font-weight: ${tokens.fontWeight_normal};
`;

export const closeBase = () => `
  border: none;
  background: none;
  padding: 0;
  display: inline-flex;
  width: 1.25rem;
  height: 1.25rem;
  align-items: center;
  justify-content: center;
  margin-left: ${tokens.spacing_100};
  border-radius: ${tokens.borderRadius_circle};
  transition-duration: 0.15s;
  transition-property: background;
`;

// This margin is a little funky, however, appears to be the best route to take to add
// spacing between icons and other tag content. Ideally, if Tag instances do not have bare
// strings inside, then the sibling combinator could be used (i.e., `> * + svg { margin-left: 2px; }`),
// however, that's not possible as bare strings do not count as sibling elements.
export const content = () => `
  & > svg {
    margin: 0 2px;
  }
`;

export const tagColor = props => {
  let bg = '';
  let border = '';
  let color = '';

  switch (props.tagColor) {
    case 'orange':
    case 'blue':
    case 'navy':
      bg = tokens.color_blue_300;
      border = tokens.color_blue_500;
      color = tokens.color_blue_800;
      break;
    case 'green':
      bg = tokens.color_green_300;
      border = tokens.color_green_500;
      color = tokens.color_green_800;
      break;
    case 'yellow':
      bg = tokens.color_yellow_200;
      border = tokens.color_yellow_400;
      color = tokens.color_yellow_700;
      break;
    case 'red':
      bg = tokens.color_red_300;
      border = tokens.color_red_500;
      color = tokens.color_red_800;
      break;
    case 'magenta':
      bg = tokens.color_magenta_300;
      border = tokens.color_magenta_500;
      color = tokens.color_magenta_800;
      break;
    case 'purple':
      bg = tokens.color_purple_300;
      border = tokens.color_purple_500;
      color = tokens.color_purple_800;
      break;
    case 'teal':
      bg = tokens.color_teal_200;
      border = tokens.color_teal_500;
      color = tokens.color_teal_800;
      break;
    case 'gray':
    default:
      bg = tokens.color_gray_300;
      border = tokens.color_gray_500;
      color = tokens.color_gray_900;
      break;
  }

  return `
    background: ${bg};
    border: 1px solid ${border};
    color: ${color};
  `;
};

export const closeColor = props => {
  let bg = '';
  let color = '';

  switch (props.tagColor) {
    case 'orange':
    case 'blue':
    case 'navy':
      bg = tokens.color_blue_500;
      color = tokens.color_blue_800;
      break;
    case 'green':
      bg = tokens.color_green_500;
      color = tokens.color_green_800;
      break;
    case 'yellow':
      bg = tokens.color_yellow_400;
      color = tokens.color_yellow_700;
      break;
    case 'red':
      bg = tokens.color_red_500;
      color = tokens.color_red_800;
      break;
    case 'magenta':
      bg = tokens.color_magenta_500;
      color = tokens.color_magenta_800;
      break;
    case 'purple':
      bg = tokens.color_purple_500;
      color = tokens.color_purple_800;
      break;
    case 'teal':
      bg = tokens.color_teal_500;
      color = tokens.color_teal_800;
      break;
    case 'gray':
    default:
      bg = tokens.color_gray_500;
      color = tokens.color_gray_900;
      break;
  }

  return `
    color: ${color};
    &:hover {
      color: ${color};
      background: ${bg};
      cursor: pointer;
    }
  `;
};
