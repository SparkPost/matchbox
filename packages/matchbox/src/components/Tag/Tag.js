import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Close } from '@sparkpost/matchbox-icons';
import { createPropTypes } from '@styled-system/prop-types';
import { margin } from 'styled-system';
import { Box } from '../Box';
import { ScreenReaderOnly } from '../ScreenReaderOnly';
import { pick } from '../../helpers/props';
import { tagBase, tagColor, closeBase, closeColor, content } from './styles';

const StyledTag = styled(Box)`
  ${tagBase}
  ${tagColor}
  ${margin}
`;

const StyledClose = styled('button')`
  ${closeBase}
  ${closeColor}
`;

const StyledContent = styled('span')`
  ${content}
`;

function Tag(props) {
  const { color, children, onRemove, className, ...rest } = props;
  const systemProps = pick(rest, margin.propNames);

  const closeMarkup = onRemove ? (
    <StyledClose onClick={onRemove} tagColor={color} type="button">
      <Close size={16} />
      <ScreenReaderOnly>Remove</ScreenReaderOnly>
    </StyledClose>
  ) : null;

  return (
    <StyledTag
      as="span"
      className={className}
      data-id={rest['data-id']}
      tagColor={color}
      hasRemove={!!onRemove}
      {...systemProps}
    >
      <StyledContent>{children}</StyledContent>
      {closeMarkup}
    </StyledTag>
  );
}

Tag.displayName = 'Tag';
Tag.propTypes = {
  /**
   * 'orange' | 'blue' | 'yellow' | 'red' | 'navy' | 'purple' | 'green' | 'magenta' | 'teal' | 'lightGray' | 'darkGray'
   */
  color: PropTypes.oneOf([
    'orange',
    'blue',
    'yellow',
    'red',
    'navy',
    'purple',
    'green',
    'magenta',
    'teal',
    'lightGray',
    'darkGray',
  ]),
  'data-id': PropTypes.string,
  /**
   * Close button is hidden unless this is provided
   */
  onRemove: PropTypes.func,
  /**
   * Tag content
   */
  children: PropTypes.node,
  ...createPropTypes(margin.propNames),
};

export default Tag;
