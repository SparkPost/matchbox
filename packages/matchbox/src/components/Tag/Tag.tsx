import React from 'react';
import styled from 'styled-components';
import { Close } from '@sparkpost/matchbox-icons';
import { margin, MarginProps } from 'styled-system';
import { Box } from '../Box';
import { ScreenReaderOnly } from '../ScreenReaderOnly';
import { pick } from '../../helpers/props';
import { tagBase, tagColor, closeBase, closeColor, content } from './styles';

type TagColorProp = {
  $tagColor?:
    | 'orange'
    | 'blue'
    | 'yellow'
    | 'red'
    | 'navy'
    | 'purple'
    | 'green'
    | 'magenta'
    | 'teal'
    | 'lightGray'
    | 'darkGray';
};

export type TagProps = MarginProps & {
  color?: TagColorProp['$tagColor'];
  className?: string;
  /**
   * Tag content
   */
  children?: React.ReactNode;
  'data-id'?: string;
  id?: string;
  /**
   * Close button is hidden unless this is provided
   */
  onRemove?: React.MouseEventHandler<HTMLButtonElement>;
};

const StyledTag = styled(Box)<TagColorProp & { $hasRemove?: boolean }>`
  ${tagBase}
  ${tagColor}
  ${margin}
`;

const StyledClose = styled.button<TagColorProp>`
  ${closeBase}
  ${closeColor}
`;

const StyledContent = styled.span`
  ${content}
`;

const Tag = React.forwardRef<HTMLSpanElement, TagProps>(function Tag(props, userRef) {
  const { color, children, onRemove, className, 'data-id': dataId, id, ...rest } = props;
  const systemProps = pick(rest, margin.propNames);

  const closeMarkup = onRemove ? (
    <StyledClose onClick={onRemove} $tagColor={color} type="button">
      <Close size={16} />
      <ScreenReaderOnly>Remove</ScreenReaderOnly>
    </StyledClose>
  ) : null;

  return (
    <StyledTag
      as="span"
      className={className}
      data-id={dataId}
      id={id}
      $tagColor={color}
      $hasRemove={!!onRemove}
      {...systemProps}
      ref={userRef}
    >
      <StyledContent>{children}</StyledContent>
      {closeMarkup}
    </StyledTag>
  );
});

Tag.displayName = 'Tag';
export default Tag;
