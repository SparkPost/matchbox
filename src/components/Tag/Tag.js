import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { Icon } from '../Icon';
import { UnstyledLink } from '../UnstyledLink';

import styles from './Tag.module.scss';

class Tag extends Component {
  static propTypes = {
    onRemove: PropTypes.func,
    children: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.node),
      PropTypes.node
    ])
  };

  render() {
    const { children, onRemove, className } = this.props;

    return (
      <div className={classnames(styles.Tag, className)}>
        <div className={styles.Content}>{ children }</div>
        <UnstyledLink
          className={styles.Close}
          onClick={onRemove}>
          <Icon name='Close'/>
        </UnstyledLink>
      </div>
    );
  }
}


export default Tag;
