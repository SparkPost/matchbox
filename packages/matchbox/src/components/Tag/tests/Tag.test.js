import React from 'react';
import { shallow } from 'enzyme';

import Tag from '../Tag';

describe('Tag', () => {
  let wrapper;
  beforeEach(() => {
    const props = {
      color: 'red'
    };

    wrapper = shallow(<Tag {...props}><span>Tag 1</span></Tag>);
  });

  it('renders without close button', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders with close button', () => {
    const fn = jest.fn();
    wrapper.setProps({ onRemove: fn, color: 'purple' });
    expect(wrapper).toMatchSnapshot();
    wrapper.find('UnstyledLink').simulate('click');
    wrapper.find('UnstyledLink').simulate('keydown', { keyCode: 32, shiftKey: false }); // Simulating spacebar, note that the `onKey` helper requires that the `shiftKey` value be explicitly defined
    expect(fn).toHaveBeenCalledTimes(2);
  });
});
