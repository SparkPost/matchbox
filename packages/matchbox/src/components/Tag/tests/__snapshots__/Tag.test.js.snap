// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag renders with close button 1`] = `
<div
  className="Tag hasRemove purple"
>
  <div
    className="Content"
  >
    <span>
      Tag 1
    </span>
  </div>
  <UnstyledLink
    className="Close"
    onClick={[MockFunction]}
    onKeyDown={[Function]}
    role="button"
    to="javascript:void(0)"
  >
    <Close
      size={16}
    />
    <ScreenReaderOnly>
      Close
    </ScreenReaderOnly>
  </UnstyledLink>
</div>
`;

exports[`Tag renders without close button 1`] = `
<div
  className="Tag red"
>
  <div
    className="Content"
  >
    <span>
      Tag 1
    </span>
  </div>
</div>
`;
